From a88a8f41489493159a20231ad10157d7c24304c5 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Thu, 6 Jun 2024 16:06:01 +0200
Subject: [PATCH 622/841] feat(stm32mp1): rework ddr supplies handling

Retrieve DDR power supplies from the device-tree.
This patch requests to describe the DDR power supplies
in the DT_DDR_COMPAT node.
This patch assumes also that the sink-source or
the bypass property was set in the device-tree in the ldo3
node.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 5276e513c58f5fbd449b629a4b4a33ce40adde44 feat(stm32mp1): rework ddr supplies handling]
Change-Id: Ida9dd705b623a8bd40dd206cc08e2d550446b172
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/388115
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
---
 .../memory-controllers/st,stm32mp1-ddr.txt    |  12 ++
 drivers/st/ddr/stm32mp1_ddr.c                 |   1 +
 plat/st/stm32mp1/plat_ddr.c                   | 121 +++++++++---------
 3 files changed, 75 insertions(+), 59 deletions(-)

diff --git a/docs/devicetree/bindings/memory-controllers/st,stm32mp1-ddr.txt b/docs/devicetree/bindings/memory-controllers/st,stm32mp1-ddr.txt
index ec610925b..310cfa94a 100644
--- a/docs/devicetree/bindings/memory-controllers/st,stm32mp1-ddr.txt
+++ b/docs/devicetree/bindings/memory-controllers/st,stm32mp1-ddr.txt
@@ -11,6 +11,18 @@ Required properties:
 		  the "ddrphyc" clock is used to check the DDR frequency
 		  at phy level according the expected value in "mem-speed" field
 
+And the supply attributes depending on the DDR type (DDR3/LPDDR2/LPDDR3).
+
+For DDR3:
+- vdd-supply	: phandle to the VDD/VDDQ regulator device tree node.
+- vref-supply	: phandle to the VREFCA/VREFDQ regulator device tree node.
+- vtt-supply	: phandle to the VTT regulator device tree node.
+
+For LPDDR2/LPDDR3:
+- vdd1-supply	: phandle to the VDD1 regulator device tree node.
+- vdd2-supply	: phandle to the VDD2/VDDCA regulator device tree node.
+- vddq-supply	: phandle to the VDDQ regulator device tree node.
+
 the next attributes are DDR parameters, they are generated by DDR tools
 included in STM32 Cube tool
 
diff --git a/drivers/st/ddr/stm32mp1_ddr.c b/drivers/st/ddr/stm32mp1_ddr.c
index c8e221879..16120ce6f 100644
--- a/drivers/st/ddr/stm32mp1_ddr.c
+++ b/drivers/st/ddr/stm32mp1_ddr.c
@@ -673,6 +673,7 @@ void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
 	}
 
 	if (ret != 0) {
+		ERROR("DDR power init failed\n");
 		panic();
 	}
 
diff --git a/plat/st/stm32mp1/plat_ddr.c b/plat/st/stm32mp1/plat_ddr.c
index 8ed5b5436..09ebe688b 100644
--- a/plat/st/stm32mp1/plat_ddr.c
+++ b/plat/st/stm32mp1/plat_ddr.c
@@ -5,6 +5,7 @@
  */
 
 #include <errno.h>
+#include <libfdt.h>
 
 #include <drivers/st/regulator.h>
 #include <drivers/st/stm32mp_ddr.h>
@@ -14,109 +15,111 @@
 static int pmic_ddr_power_init(enum ddr_type ddr_type)
 {
 	int status;
-	uint16_t buck3_min_mv __unused;
-	struct rdev *buck2, *buck3 __unused, *vref;
-	struct rdev *ldo3 __unused;
+	void *fdt;
+	int node;
 
-	buck2 = regulator_get_by_name("buck2");
-	if (buck2 == NULL) {
-		return -ENOENT;
+	if (fdt_get_address(&fdt) == 0) {
+		return -FDT_ERR_NOTFOUND;
 	}
 
-#if STM32MP15
-	ldo3 = regulator_get_by_name("ldo3");
-	if (ldo3 == NULL) {
-		return -ENOENT;
-	}
-#endif
-
-	vref = regulator_get_by_name("vref_ddr");
-	if (vref == NULL) {
-		return -ENOENT;
+	node = fdt_node_offset_by_compatible(fdt, -1, DT_DDR_COMPAT);
+	if (node < 0) {
+		ERROR("%s: Cannot read DDR node in DT\n", __func__);
+		return -EINVAL;
 	}
 
 	switch (ddr_type) {
-	case STM32MP_DDR3:
-#if STM32MP15
-		status = regulator_set_flag(ldo3, REGUL_SINK_SOURCE);
-		if (status != 0) {
-			return status;
+	case STM32MP_DDR3: {
+		struct rdev *vref;
+		struct rdev *vdd;
+		struct rdev *vtt;
+
+		vdd = regulator_get_by_supply_name(fdt, node, "vdd");
+		if (vdd == NULL) {
+			return -ENOENT;
 		}
-#endif
 
-		status = regulator_set_min_voltage(buck2);
+		status = regulator_set_min_voltage(vdd);
 		if (status != 0) {
 			return status;
 		}
 
-		status = regulator_enable(buck2);
+
+		status = regulator_enable(vdd);
 		if (status != 0) {
 			return status;
 		}
 
-		status = regulator_enable(vref);
-		if (status != 0) {
-			return status;
+		vref = regulator_get_by_supply_name(fdt, node, "vref");
+		if (vref != NULL) {
+			status = regulator_enable(vref);
+			if (status != 0) {
+				return status;
+			}
 		}
 
-#if STM32MP15
-		status = regulator_enable(ldo3);
-		if (status != 0) {
-			return status;
+		vtt = regulator_get_by_supply_name(fdt, node, "vtt");
+		if (vtt != NULL) {
+			status = regulator_enable(vtt);
+			if (status != 0) {
+				return status;
+			}
 		}
-#endif
 		break;
+	}
 
 	case STM32MP_LPDDR2:
-	case STM32MP_LPDDR3:
-#if STM32MP15
-		/*
-		 * Set LDO3 to 1.8V according  BUCK3 voltage
-		 * => bypass mode if BUCK3 = 1.8V
-		 * => normal mode if BUCK3 != 1.8V
-		 */
-		buck3 = regulator_get_by_name("buck3");
-		if (buck3 == NULL) {
+	case STM32MP_LPDDR3: {
+		struct rdev *vddq;
+		struct rdev *vdd1_ddr;
+		struct rdev *vdd2_ddr;
+
+		vdd1_ddr = regulator_get_by_supply_name(fdt, node, "vdd1");
+		if (vdd1_ddr == NULL) {
 			return -ENOENT;
 		}
 
-		regulator_get_range(buck3, &buck3_min_mv, NULL);
+		vdd2_ddr = regulator_get_by_supply_name(fdt, node, "vdd2");
+		if (vdd2_ddr == NULL) {
+			return -ENOENT;
+		}
 
-		if (buck3_min_mv != 1800) {
-			status = regulator_set_min_voltage(ldo3);
-			if (status != 0) {
-				return status;
-			}
-		} else {
-			status = regulator_set_flag(ldo3, REGUL_ENABLE_BYPASS);
-			if (status != 0) {
-				return status;
-			}
+		vddq = regulator_get_by_supply_name(fdt, node, "vddq");
+		if (vddq == NULL) {
+			return -ENOENT;
 		}
-#endif
 
-		status = regulator_set_min_voltage(buck2);
+		status = regulator_set_min_voltage(vdd1_ddr);
 		if (status != 0) {
 			return status;
 		}
 
-#if STM32MP15
-		status = regulator_enable(ldo3);
+		status = regulator_set_min_voltage(vdd2_ddr);
 		if (status != 0) {
 			return status;
 		}
-#endif
 
-		status = regulator_enable(buck2);
+		status = regulator_set_min_voltage(vddq);
 		if (status != 0) {
 			return status;
 		}
 
-		status = regulator_enable(vref);
+		status = regulator_enable(vdd1_ddr);
+		if (status != 0) {
+			return status;
+		}
+
+		status = regulator_enable(vdd2_ddr);
+		if (status != 0) {
+			return status;
+		}
+
+		status = regulator_enable(vddq);
 		if (status != 0) {
 			return status;
 		}
 		break;
+	}
 
 	default:
 		break;
-- 
2.47.2

